tests/test_BT_piecewise.py

```diff
@@ -0,0 +1,560 @@
+        np.unique(index, return_counts=True)[1][0],
+        np.unique(index, return_counts=True)[1][1],
+    ]
+    expected_toas_in_each_group = [10, 10]
+    is_there_ten_toas_per_group = np.array_equal(
+        should_be_ten_toas_in_each_group, expected_toas_in_each_group
+    )
+    assert is_there_ten_toas_per_group
+
+
+@pytest.mark.parametrize("param", ["A1X", "T0X"])
+def test_paramX_per_toa_matches_corresponding_model_value(
+    param, build_piecewise_model_with_two_pieces, make_toas_to_go_with_two_piece_model
+):
+    # Testing the correct piecewise parameters are being assigned to each toa.
+    # Operates on the piecewise_parameter_from_information_array function. Requires group_index fn to be run so we have an array of length(ntoas), filled with information on which group a toa belongs to.
+    # Uses this array to apply T0X_i/A1X_i to corresponding indexes from group_index fn call. i.e. for T0X_i,T0X_j,T0X_k values and group_index return: [i,j,k] the output would be [T0X_i,T0X_j,T0X_k]
+    m_piecewise = build_piecewise_model_with_two_pieces
+    toa = make_toas_to_go_with_two_piece_model
+    m_piecewise.setup()
+    rs = pint.residuals.Residuals(toa, m_piecewise)
+    if param == "A1X":
+
+        paramX_per_toa = m_piecewise.get_A1Xs_associated_with_toas(toa)
+        test_val = [m_piecewise.A1X_0000.quantity, m_piecewise.A1X_0001.quantity]
+        should_toa_reference_piecewise_parameter = [
+            m_piecewise.does_toa_reference_piecewise_parameter(toa, "A1X_0000"),
+            m_piecewise.does_toa_reference_piecewise_parameter(toa, "A1X_0001"),
+        ]
+    elif param == "T0X":
+        paramX_per_toa = m_piecewise.get_T0Xs_associated_with_toas(toa)
+        test_val = [m_piecewise.T0X_0000.quantity, m_piecewise.T0X_0001.quantity]
+        should_toa_reference_piecewise_parameter = [
+            m_piecewise.does_toa_reference_piecewise_parameter(toa, "T0X_0000"),
+            m_piecewise.does_toa_reference_piecewise_parameter(toa, "T0X_0001"),
+        ]
+
+    do_toas_reference_first_piecewise_parameter = np.isclose(
+        (paramX_per_toa.value - test_val[0].value), 0, atol=1e-6, rtol=0
+    )
+    do_toas_reference_second_piecewise_parameter = np.isclose(
+        (paramX_per_toa.value - test_val[1].value), 0, atol=1e-6, rtol=0
+    )
+    do_toas_reference_piecewise_parameter = [
+        do_toas_reference_first_piecewise_parameter,
+        do_toas_reference_second_piecewise_parameter,
+    ]
+    do_arrays_match = np.array_equal(
+        do_toas_reference_piecewise_parameter, should_toa_reference_piecewise_parameter
+    )
+    assert do_arrays_match
+
+
+def test_problematic_group_indexes_and_ranges(build_model):
+    # Test to flag issues with problematic group indexes
+    m_piecewise = build_model
+    with pytest.raises(Exception):
+        assert m_piecewise.add_group_range(
+            m_piecewise.START.value, m_piecewise.FINISH.value, j=-1
+        )
+        assert m_piecewise.add_group_range(
```

Comment:
This needs to be in a separate `with` statement - the first exception exits the `with`.

Reply:
Fixed to only include a single assertion in a `with` statement. I realise most of these have been removed during the update